File: skills/writing-tests-that-can-fail/evals/recorded/2026-09-25/d-account-js/skill-haiku-3/fixture_d_account_test.mjs
import test from "node:test";
import assert from "node:assert";
import { AccountService } from "./fixture_d_account.mjs";

// Fake repo: in-memory implementation of AccountRepo
class FakeAccountRepo {
  constructor() {
    this.accounts = new Map();
    this.statusUpdates = [];
  }

  get(userId) {
    if (!this.accounts.has(userId)) {
      throw new Error(`Account not found for user ${userId}`);
    }
    return this.accounts.get(userId);
  }

  setStatus(userId, status) {
    this.statusUpdates.push({ userId, status });
    const account = this.accounts.get(userId);
    if (account) {
      account.status = status;
    }
  }

  addAccount(userId, email, status = "active") {
    this.accounts.set(userId, { email, status });
  }
}

// Fake mailer: in-memory implementation of Mailer
class FakeMailer {
  constructor() {
    this.sentEmails = [];
  }

  sendGoodbye(email) {
    this.sentEmails.push(email);
  }
}

test("AccountService.deactivate: active account transitions to inactive", () => {
  const repo = new FakeAccountRepo();
  const mailer = new FakeMailer();
  const service = new AccountService(repo, mailer);

  repo.addAccount("user123", "alice@example.com", "active");
  service.deactivate("user123");

  const account = repo.get("user123");
  assert.strictEqual(account.status, "inactive", "Account status should be set to inactive");
});

test("AccountService.deactivate: sends goodbye email to account email address", () => {
  const repo = new FakeAccountRepo();
  const mailer = new FakeMailer();
  const service = new AccountService(repo, mailer);

  repo.addAccount("user456", "bob@example.com", "active");
  service.deactivate("user456");

  assert.deepStrictEqual(
    mailer.sentEmails,
    ["bob@example.com"],
    "Goodbye email should be sent to the account's email address"
  );
});

test("AccountService.deactivate: already inactive account sends no email", () => {
  const repo = new FakeAccountRepo();
  const mailer = new FakeMailer();
  const service = new AccountService(repo, mailer);

  repo.addAccount("user789", "charlie@example.com", "inactive");
  service.deactivate("user789");

  assert.deepStrictEqual(
    mailer.sentEmails,
    [],
    "No email should be sent when account is already inactive"
  );
});

test("AccountService.deactivate: already inactive account does not call setStatus", () => {
  const repo = new FakeAccountRepo();
  const mailer = new FakeMailer();
  const service = new AccountService(repo, mailer);

  repo.addAccount("user999", "dave@example.com", "inactive");
  service.deactivate("user999");

  assert.deepStrictEqual(
    repo.statusUpdates,
    [],
    "setStatus should not be called for already inactive accounts"
  );
});

test("AccountService.deactivate: active account calls setStatus with inactive", () => {
  const repo = new FakeAccountRepo();
  const mailer = new FakeMailer();
  const service = new AccountService(repo, mailer);

  repo.addAccount("user111", "eve@example.com", "active");
  service.deactivate("user111");

  assert.deepStrictEqual(
    repo.statusUpdates,
    [{ userId: "user111", status: "inactive" }],
    "setStatus should be called with correct userId and 'inactive' status"
  );
});

test("AccountService.deactivate: verifies early return prevents double deactivation side effects", () => {
  const repo = new FakeAccountRepo();
  const mailer = new FakeMailer();
  const service = new AccountService(repo, mailer);

  repo.addAccount("user222", "frank@example.com", "inactive");

  // Deactivate an already-inactive account
  service.deactivate("user222");

  // Status should remain unchanged
  const account = repo.get("user222");
  assert.strictEqual(account.status, "inactive", "Status should remain inactive");

  // Email list should be empty (no goodbye sent to already-inactive account)
  assert.strictEqual(mailer.sentEmails.length, 0, "No email should be sent for already-inactive account");
});

File: evals/writing-tests-that-can-fail/recorded/2026-09-25/d-account-py/skill-haiku-1/fixture_d_account_test.py
"""Tests for AccountService deactivate logic."""

import pytest
from fixture_d_account import Account, AccountRepo, Mailer, AccountService


class FakeAccountRepo(AccountRepo):
    def __init__(self):
        self.accounts = {}
        self.status_updates = []

    def get(self, user_id: str) -> Account:
        if user_id not in self.accounts:
            raise ValueError(f"Account not found: {user_id}")
        return self.accounts[user_id]

    def set_status(self, user_id: str, status: str) -> None:
        self.status_updates.append((user_id, status))
        if user_id in self.accounts:
            self.accounts[user_id].status = status


class FakeMailer(Mailer):
    def __init__(self):
        self.messages = []

    def send_goodbye(self, email: str) -> None:
        self.messages.append(email)


@pytest.fixture
def repo():
    return FakeAccountRepo()


@pytest.fixture
def mailer():
    return FakeMailer()


@pytest.fixture
def service(repo, mailer):
    return AccountService(repo, mailer)


class TestAccountServiceDeactivate:
    def test_deactivate_active_account_sets_status_to_inactive(self, repo, service):
        repo.accounts["user123"] = Account(user_id="user123", email="test@example.com", status="active")

        service.deactivate("user123")

        assert repo.accounts["user123"].status == "inactive"

    def test_deactivate_active_account_sends_goodbye_email(self, repo, mailer, service):
        repo.accounts["user123"] = Account(user_id="user123", email="alice@example.com", status="active")

        service.deactivate("user123")

        assert "alice@example.com" in mailer.messages

    def test_deactivate_active_account_calls_repo_set_status(self, repo, service):
        repo.accounts["user123"] = Account(user_id="user123", email="test@example.com", status="active")

        service.deactivate("user123")

        assert ("user123", "inactive") in repo.status_updates

    def test_deactivate_inactive_account_is_noop(self, repo, mailer, service):
        repo.accounts["user123"] = Account(user_id="user123", email="test@example.com", status="inactive")

        service.deactivate("user123")

        assert repo.status_updates == []
        assert mailer.messages == []

    def test_deactivate_sends_email_to_correct_recipient(self, repo, mailer, service):
        repo.accounts["user456"] = Account(user_id="user456", email="bob@company.org", status="active")

        service.deactivate("user456")

        assert mailer.messages == ["bob@company.org"]

    def test_deactivate_multiple_accounts_independently(self, repo, mailer, service):
        repo.accounts["user1"] = Account(user_id="user1", email="alice@example.com", status="active")
        repo.accounts["user2"] = Account(user_id="user2", email="bob@example.com", status="active")

        service.deactivate("user1")
        service.deactivate("user2")

        assert repo.accounts["user1"].status == "inactive"
        assert repo.accounts["user2"].status == "inactive"
        assert mailer.messages == ["alice@example.com", "bob@example.com"]

    def test_deactivate_already_inactive_twice_stays_noop(self, repo, mailer, service):
        repo.accounts["user123"] = Account(user_id="user123", email="test@example.com", status="inactive")

        service.deactivate("user123")
        service.deactivate("user123")

        assert repo.status_updates == []
        assert mailer.messages == []

    def test_deactivate_active_then_deactivate_again_no_second_email(self, repo, mailer, service):
        repo.accounts["user123"] = Account(user_id="user123", email="test@example.com", status="active")

        service.deactivate("user123")
        service.deactivate("user123")

        assert mailer.messages == ["test@example.com"]
        assert len(repo.status_updates) == 1
